File: sdks/js/packages/spark-sdk/src/tests/utils/test-faucet.ts
import { bytesToHex, hexToBytes } from "@noble/curves/abstract/utils";
import { schnorr, secp256k1 } from "@noble/curves/secp256k1";
import * as btc from "@scure/btc-signer";
import { Address, OutScript, SigHash, Transaction } from "@scure/btc-signer";
import { TransactionInput, TransactionOutput } from "@scure/btc-signer/psbt";
import { taprootTweakPrivKey } from "@scure/btc-signer/utils";
import { RPCError } from "../../errors/index.js";
import {
  getP2TRAddressFromPublicKey,
  getP2TRScriptFromPublicKey,
} from "../../utils/bitcoin.js";
import { getNetwork, Network } from "../../utils/network.js";

// Static keys for deterministic testing
// P2TRAddress: bcrt1p2uy9zw5ltayucsuzl4tet6ckelzawp08qrtunacscsszflye907q62uqhl
const STATIC_FAUCET_KEY = hexToBytes(
  "deadbeef1337cafe4242424242424242deadbeef1337cafe4242424242424242",
);

// P2TRAddress: bcrt1pwr5k38p68ceyrnm2tvrp50dvmg3grh6uvayjl3urwtxejhd3dw4swz6p58
const STATIC_MINING_KEY = hexToBytes(
  "1337cafe4242deadbeef4242424242421337cafe4242deadbeef424242424242",
);
const SATS_PER_BTC = 100_000_000;

export type FaucetCoin = {
  key: Uint8Array;
  outpoint: TransactionInput;
  txout: TransactionOutput;
};

// The amount of satoshis to put in each faucet coin to be used in tests
const COIN_AMOUNT = 10_000_000n;
const FEE_AMOUNT = 1000n;
const TARGET_NUM_COINS = 20;

export class BitcoinFaucet {
  private coins: FaucetCoin[] = [];
  private static instance: BitcoinFaucet | null = null;
  private miningAddress: string;
  private lock: Promise<void> = Promise.resolve();

  private constructor(
    private url: string = "http://127.0.0.1:8332",
    private username: string = "testutil",
    private password: string = "testutilpassword",
  ) {
    this.miningAddress = getP2TRAddressFromPublicKey(
      secp256k1.getPublicKey(STATIC_MINING_KEY),
      Network.LOCAL,
    );
  }

  static getInstance(
    url: string = "http://127.0.0.1:8332",
    username: string = "testutil",
    password: string = "testutilpassword",
  ): BitcoinFaucet {
    if (!BitcoinFaucet.instance) {
      BitcoinFaucet.instance = new BitcoinFaucet(url, username, password);
    }
    return BitcoinFaucet.instance;
  }

  private async withLock<T>(operation: () => Promise<T>): Promise<T> {
    const current = this.lock;
    let resolve: () => void;
    this.lock = new Promise<void>((r) => (resolve = r));
    await current;
    try {
      return await operation();
    } finally {
      resolve!();
    }
  }

  async fund(): Promise<FaucetCoin> {
    return this.withLock(async () => {
      if (this.coins.length === 0) {
        await this.refill();
      }

      const coin = this.coins[0];
      if (!coin) {
        throw new Error("Failed to get coin from faucet");
      }
      this.coins = this.coins.slice(1);
      return coin;
    });
  }

  private async refill(): Promise<void> {
    const minerPubKey = secp256k1.getPublicKey(STATIC_MINING_KEY);
    const address = getP2TRAddressFromPublicKey(minerPubKey, Network.LOCAL);

    // Use scantxoutset to find UTXOs
    const scanResult = await this.call("scantxoutset", [
      "start",
      [`addr(${address})`],
    ]);

    let selectedUtxo;
    let selectedUtxoAmountSats;
    if (!scanResult.success || scanResult.unspents.length === 0) {
      const blockHash = await this.generateToAddress(1, address);
      const block = await this.getBlock(blockHash[0]);
      const fundingTx = Transaction.fromRaw(hexToBytes(block.tx[0].hex), {
        allowUnknownOutputs: true,
      });

      await this.generateToAddress(100, this.miningAddress);

      selectedUtxo = {
        txid: block.tx[0].txid,
        vout: 0,
        amount: fundingTx.getOutput(0)!.amount!, // Already in sats
      };
      selectedUtxoAmountSats = BigInt(selectedUtxo.amount);
    } else {
      selectedUtxo = scanResult.unspents.find((utxo) => {
        const isValueEnough =
          BigInt(Math.floor(utxo.amount * SATS_PER_BTC)) >=
          COIN_AMOUNT + FEE_AMOUNT;
        const isMature = scanResult.height - utxo.height >= 100;
        return isValueEnough && isMature;
      });

      if (!selectedUtxo) {
        throw new Error("No UTXO large enough to create even one faucet coin");
      }
      selectedUtxoAmountSats = BigInt(
        Math.floor(selectedUtxo.amount * SATS_PER_BTC),
      );
    }

    const maxPossibleCoins = Number(
      (selectedUtxoAmountSats - FEE_AMOUNT) / COIN_AMOUNT,
    );
    const numCoinsToCreate = Math.min(maxPossibleCoins, TARGET_NUM_COINS);

    if (numCoinsToCreate < 1) {
      throw new Error(
        `Selected UTXO (${selectedUtxoAmountSats} sats) is too small to create even one faucet coin of ${COIN_AMOUNT} sats`,
      );
    }

    const splitTx = new Transaction();
    splitTx.addInput({
      txid: selectedUtxo.txid,
      index: selectedUtxo.vout,
    });

    const faucetPubKey = secp256k1.getPublicKey(STATIC_FAUCET_KEY);
    const script = getP2TRScriptFromPublicKey(faucetPubKey, Network.LOCAL);
    for (let i = 0; i < numCoinsToCreate; i++) {
      splitTx.addOutput({
        script,
        amount: COIN_AMOUNT,
      });
    }

    const remainingValue =
      selectedUtxoAmountSats -
      COIN_AMOUNT * BigInt(numCoinsToCreate) -
      FEE_AMOUNT;
    const minerScript = getP2TRScriptFromPublicKey(minerPubKey, Network.LOCAL);
    if (remainingValue > 0n) {
      splitTx.addOutput({
        script: minerScript,
        amount: remainingValue,
      });
    }

    const signedSplitTx = await this.signFaucetCoin(
      splitTx,
      {
        amount: selectedUtxoAmountSats,
        script: minerScript,
      },
      STATIC_MINING_KEY,
    );

    await this.broadcastTx(bytesToHex(signedSplitTx.extract()));

    const splitTxId = signedSplitTx.id;
    for (let i = 0; i < numCoinsToCreate; i++) {
      this.coins.push({
        key: STATIC_FAUCET_KEY,
        outpoint: {
          txid: hexToBytes(splitTxId),
          index: i,
        },
        txout: signedSplitTx.getOutput(i)!,
      });
    }
  }

  async sendFaucetCoinToP2WPKHAddress(pubKey: Uint8Array) {
    const sendToPubKeyTx = new Transaction();

    // For P2WPKH, we need to hash the public key

    // Create a P2WPKH address
    const p2wpkhAddress = btc.p2wpkh(pubKey, getNetwork(Network.LOCAL)).address;
    if (!p2wpkhAddress) {
      throw new Error("Invalid P2WPKH address");
    }

    // Get the coin to spend
    const coinToSend = await this.fund();
    if (!coinToSend) {
      throw new Error("No coins available");
    }

    // Add the input
    sendToPubKeyTx.addInput(coinToSend.outpoint);

    // Add the output using the address directly, but subtract FEE_AMOUNT to ensure there's a fee
    sendToPubKeyTx.addOutputAddress(
      p2wpkhAddress,
      COIN_AMOUNT - FEE_AMOUNT,
      getNetwork(Network.LOCAL),
    );

    // Sign the transaction and get the signed result
    const signedTx = await this.signFaucetCoin(
      sendToPubKeyTx,
      coinToSend.txout,
      coinToSend.key,
    );

    // Broadcast the signed transaction
    await this.broadcastTx(bytesToHex(signedTx.extract()));
  }

  async signFaucetCoin(
    unsignedTx: Transaction,
    fundingTxOut: TransactionOutput,
    key: Uint8Array,
  ): Promise<Transaction> {
    const pubKey = secp256k1.getPublicKey(key);
    const internalKey = pubKey.slice(1); // Remove the 0x02/0x03 prefix

    const script = getP2TRScriptFromPublicKey(pubKey, Network.LOCAL);

    unsignedTx.updateInput(0, {
      tapInternalKey: internalKey,
      witnessUtxo: {
        script,
        amount: fundingTxOut.amount!,
      },
    });

    const sighash = unsignedTx.preimageWitnessV1(
      0,
      new Array(unsignedTx.inputsLength).fill(script),
      SigHash.DEFAULT,
      new Array(unsignedTx.inputsLength).fill(fundingTxOut.amount!),
    );

    const merkleRoot = new Uint8Array();
    const tweakedKey = taprootTweakPrivKey(key, merkleRoot);
    if (!tweakedKey)
      throw new Error("Invalid private key for taproot tweaking");

    const signature = schnorr.sign(sighash, tweakedKey);

    unsignedTx.updateInput(0, {
      tapKeySig: signature,
    });

    unsignedTx.finalize();

    return unsignedTx;
  }

  // MineBlocks mines the specified number of blocks to a random address
  // and returns the block hashes.
  async mineBlocks(numBlocks: number) {
    return await this.generateToAddress(numBlocks, this.miningAddress);
  }

  private async call(method: string, params: any[]) {
    try {
      const response = await fetch(this.url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: "Basic " + btoa(`${this.username}:${this.password}`),
        },
        body: JSON.stringify({
          jsonrpc: "1.0",
          id: "spark-js",
          method,
          params,
        }),
      });

      const data = await response.json();
      if (data.error) {
        console.error(`RPC Error for method ${method}:`, data.error);
        throw new RPCError(`Bitcoin RPC error: ${data.error.message}`, {
          method,
          params,
          code: data.error.code,
        });
      }

      return data.result;
    } catch (error) {
      if (error instanceof RPCError) {
        throw error;
      }
      throw new RPCError(
        "Failed to call Bitcoin RPC",
        {
          method,
          params,
        },
        error as Error,
      );
    }
  }

  async generateToAddress(numBlocks: number, address: string) {
    return await this.call("generatetoaddress", [numBlocks, address]);
  }

  async getBlock(blockHash: string) {
    return await this.call("getblock", [blockHash, 2]);
  }

  async broadcastTx(txHex: string) {
    let response = await this.call("sendrawtransaction", [txHex, 0]);
    return response;
  }

  async getNewAddress(): Promise<string> {
    const key = secp256k1.utils.randomPrivateKey();
    const pubKey = secp256k1.getPublicKey(key);
    return getP2TRAddressFromPublicKey(pubKey, Network.LOCAL);
  }

  async sendToAddress(address: string, amount: bigint): Promise<Transaction> {
    const coin = await this.fund();
    if (!coin) {
      throw new Error("No coins available");
    }

    const tx = new Transaction();
    tx.addInput(coin.outpoint);

    const availableAmount = COIN_AMOUNT - FEE_AMOUNT;

    const destinationAddress = Address(getNetwork(Network.LOCAL)).decode(
      address,
    );
    const destinationScript = OutScript.encode(destinationAddress);
    tx.addOutput({
      script: destinationScript,
      amount: amount,
    });

    const changeAmount = availableAmount - amount;
    if (changeAmount > 0) {
      const changeKey = secp256k1.utils.randomPrivateKey();
      const changePubKey = secp256k1.getPublicKey(changeKey);
      const changeScript = getP2TRScriptFromPublicKey(
        changePubKey,
        Network.LOCAL,
      );
      tx.addOutput({
        script: changeScript,
        amount: changeAmount,
      });
    }

    const signedTx = await this.signFaucetCoin(tx, coin.txout, coin.key);
    const txHex = bytesToHex(signedTx.extract());
    await this.broadcastTx(txHex);

    await this.generateToAddress(1, address);

    return signedTx;
  }

  async getRawTransaction(txid: string) {
    return await this.call("getrawtransaction", [txid, 2]);
  }
}

File: spark/test_util/test_faucet.go
package testutil

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"sync"

	"github.com/btcsuite/btcd/btcutil"
	"github.com/btcsuite/btcd/chaincfg/chainhash"
	"github.com/btcsuite/btcd/rpcclient"
	"github.com/btcsuite/btcd/txscript"
	"github.com/btcsuite/btcd/wire"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/lightsparkdev/spark/common"
)

var (
	// Static keys for deterministic testing
	// P2TRAddress: bcrt1p2uy9zw5ltayucsuzl4tet6ckelzawp08qrtunacscsszflye907q62uqhl
	staticFaucetKeyBytes, _ = hex.DecodeString("deadbeef1337cafe4242424242424242deadbeef1337cafe4242424242424242")
	staticFaucetKey         = secp256k1.PrivKeyFromBytes(staticFaucetKeyBytes)

	// P2TRAddress: bcrt1pwr5k38p68ceyrnm2tvrp50dvmg3grh6uvayjl3urwtxejhd3dw4swz6p58
	staticMiningKeyBytes, _ = hex.DecodeString("1337cafe4242deadbeef4242424242421337cafe4242deadbeef424242424242")
	staticMiningKey         = secp256k1.PrivKeyFromBytes(staticMiningKeyBytes)

	// Constants for coin amounts
	coinAmountSats int64 = 10_000_000
	feeAmountSats  int64 = 1_000
	targetNumCoins       = 20

	// Singleton instance
	instance *Faucet
	once     sync.Once
)

// scanUnspent represents an unspent output found by scanning
type scanUnspent struct {
	TxID   string      `json:"txid"`
	Vout   uint32      `json:"vout"`
	Amount json.Number `json:"amount"`
	Height int64       `json:"height"`
}

// scanTxOutSetResult represents the result of scanning the UTXO set
type scanTxOutSetResult struct {
	Success  bool          `json:"success"`
	Height   int64         `json:"height"`
	Unspents []scanUnspent `json:"unspents"`
}

type uTXO struct {
	TxID   string
	Vout   uint32
	Amount int64
	Height int64
}

// NewRegtestClient returns a new rpcclient.Client with a hard-coded
// config for our integration tests.
func NewRegtestClient() (*rpcclient.Client, error) {
	connConfig := rpcclient.ConnConfig{
		Host:         "127.0.0.1:8332",
		User:         "testutil",
		Pass:         "testutilpassword",
		Params:       "regtest",
		DisableTLS:   true,
		HTTPPostMode: true,
	}
	return rpcclient.New(
		&connConfig,
		nil,
	)
}

type FaucetCoin struct {
	Key      *secp256k1.PrivateKey
	OutPoint *wire.OutPoint
	TxOut    *wire.TxOut
}

type Faucet struct {
	client  *rpcclient.Client
	coinsMu sync.Mutex
	coins   []FaucetCoin
}

// GetFaucetInstance returns the singleton instance of the Faucet
func GetFaucetInstance(client *rpcclient.Client) *Faucet {
	once.Do(func() {
		instance = &Faucet{
			client:  client,
			coinsMu: sync.Mutex{},
			coins:   make([]FaucetCoin, 0),
		}
	})
	return instance
}

// Fund returns a faucet coin, which is a UTXO that can be spent in a test.
func (f *Faucet) Fund() (FaucetCoin, error) {
	if len(f.coins) == 0 {
		err := f.Refill()
		if err != nil {
			return FaucetCoin{}, err
		}
	}
	f.coinsMu.Lock()
	defer f.coinsMu.Unlock()
	coin := f.coins[0]
	f.coins = f.coins[1:]
	return coin, nil
}

// btcToSats converts a BTC amount (as a decimal string) to satoshis
func btcToSats(btc json.Number) (int64, error) {
	f, err := btc.Float64()
	if err != nil {
		return 0, err
	}
	amount, err := btcutil.NewAmount(f)
	if err != nil {
		return 0, err
	}
	return int64(amount), nil
}

// scanForSpendableUTXOs scans for any spendable UTXOs at the mining address
func (f *Faucet) scanForSpendableUTXOs() ([]uTXO, int64, error) {
	miningPubKey := staticMiningKey.PubKey()
	miningAddress, err := common.P2TRRawAddressFromPublicKey(miningPubKey.SerializeCompressed(), common.Regtest)
	if err != nil {
		return nil, 0, err
	}

	descriptor := fmt.Sprintf("addr(%s)", miningAddress)
	params := []json.RawMessage{
		json.RawMessage(`"start"`),
		json.RawMessage(fmt.Sprintf(`["%s"]`, descriptor)),
	}

	result, err := f.client.RawRequest("scantxoutset", params)
	if err != nil {
		return nil, 0, err
	}

	var scanResult scanTxOutSetResult
	err = json.Unmarshal(result, &scanResult)
	if err != nil {
		return nil, 0, err
	}

	if !scanResult.Success {
		return nil, scanResult.Height, fmt.Errorf("scan failed")
	}

	var utxos []uTXO
	for _, unspent := range scanResult.Unspents {
		sats, err := btcToSats(unspent.Amount)
		if err != nil {
			continue
		}
		utxos = append(utxos, uTXO{
			TxID:   unspent.TxID,
			Vout:   unspent.Vout,
			Amount: sats,
			Height: unspent.Height,
		})
	}

	return utxos, scanResult.Height, nil
}

// findSuitableUTXO finds a UTXO that is large enough and mature enough to use
func (f *Faucet) findSuitableUTXO() (*uTXO, error) {
	utxos, height, err := f.scanForSpendableUTXOs()
	if err != nil {
		return nil, err
	}

	minAmount := coinAmountSats + feeAmountSats
	for _, utxo := range utxos {
		isMature := height-utxo.Height >= 100
		isValueEnough := utxo.Amount >= minAmount

		if isMature && isValueEnough {
			return &utxo, nil
		}
	}

	return nil, nil
}

// Refill mines a block to the faucet if needed, then crafts a new transaction to split it
// into a bunch outputs (coins), which are then freely given away for various tests to use.
func (f *Faucet) Refill() error {
	f.coinsMu.Lock()
	defer f.coinsMu.Unlock()

	selectedUTXO, err := f.findSuitableUTXO()
	if err != nil {
		return err
	}

	var fundingTx *wire.MsgTx
	var fundingTxOut *wire.TxOut

	if selectedUTXO != nil {
		txHash, err := chainhash.NewHashFromStr(selectedUTXO.TxID)
		if err != nil {
			return err
		}
		tx, err := f.client.GetRawTransaction(txHash)
		if err != nil {
			return err
		}
		fundingTx = tx.MsgTx()
		fundingTxOut = fundingTx.TxOut[selectedUTXO.Vout]
	} else {
		// No suitable UTXO found, need to mine a new block
		miningPubKey := staticMiningKey.PubKey()
		miningAddress, err := common.P2TRRawAddressFromPublicKey(miningPubKey.SerializeCompressed(), common.Regtest)
		if err != nil {
			return err
		}

		blockHash, err := f.client.GenerateToAddress(1, miningAddress, nil)
		if err != nil {
			return err
		}

		block, err := f.client.GetBlockVerboseTx(blockHash[0])
		if err != nil {
			return err
		}
		fundingTx = wire.NewMsgTx(2)
		txBytes, err := hex.DecodeString(block.Tx[0].Hex)
		if err != nil {
			return err
		}
		err = fundingTx.Deserialize(bytes.NewReader(txBytes))
		if err != nil {
			return err
		}
		fundingTxOut = fundingTx.TxOut[0]

		// Mine 100 blocks to make funds spendable
		_, err = f.client.GenerateToAddress(100, miningAddress, nil)
		if err != nil {
			return err
		}
	}

	splitTx := wire.NewMsgTx(2)
	var fundingOutPoint *wire.OutPoint
	if selectedUTXO != nil {
		txHash, err := chainhash.NewHashFromStr(selectedUTXO.TxID)
		if err != nil {
			return err
		}
		fundingOutPoint = wire.NewOutPoint(txHash, selectedUTXO.Vout)
	} else {
		fundingTxid := fundingTx.TxHash()
		fundingOutPoint = wire.NewOutPoint(&fundingTxid, 0)
	}
	splitTx.AddTxIn(wire.NewTxIn(fundingOutPoint, nil, nil))

	initialValueSats := fundingTxOut.Value
	maxPossibleCoins := (initialValueSats - feeAmountSats) / coinAmountSats
	numCoinsToCreate := min(int64(targetNumCoins), maxPossibleCoins)

	if numCoinsToCreate < 1 {
		log.Printf("Selected UTXO (%d sats) is too small to create even one faucet coin of %d sats", initialValueSats, coinAmountSats)
		return nil
	}

	faucetPubKey := staticFaucetKey.PubKey()
	faucetScript, err := common.P2TRScriptFromPubKey(faucetPubKey)
	if err != nil {
		return err
	}

	for i := int64(0); i < numCoinsToCreate; i++ {
		splitTx.AddTxOut(wire.NewTxOut(coinAmountSats, faucetScript))
	}

	remainingValue := initialValueSats - (numCoinsToCreate * coinAmountSats) - feeAmountSats
	if remainingValue > 0 {
		miningScript, err := common.P2TRScriptFromPubKey(staticMiningKey.PubKey())
		if err != nil {
			return err
		}
		splitTx.AddTxOut(wire.NewTxOut(remainingValue, miningScript))
	}

	signedSplitTx, err := SignFaucetCoin(splitTx, fundingTxOut, staticMiningKey)
	if err != nil {
		return err
	}
	_, err = f.client.SendRawTransaction(signedSplitTx, true)
	if err != nil {
		return err
	}

	splitTxid := signedSplitTx.TxHash()
	for i := 0; i < int(numCoinsToCreate); i++ {
		faucetCoin := FaucetCoin{
			Key:      staticFaucetKey,
			OutPoint: wire.NewOutPoint(&splitTxid, uint32(i)),
			TxOut:    signedSplitTx.TxOut[i],
		}
		f.coins = append(f.coins, faucetCoin)
	}
	log.Printf("Refilled faucet with %d coins", len(f.coins))

	return nil
}

// SignFaucetCoin signs the first input of the given transaction with the given key,
// and returns the signed transaction. Note this expects to be spending
// a taproot output, with the spendingTxOut and key coming from a FaucetCoin from `faucet.Fund()`.
func SignFaucetCoin(unsignedTx *wire.MsgTx, spendingTxOut *wire.TxOut, key *secp256k1.PrivateKey) (*wire.MsgTx, error) {
	prevOutputFetcher := txscript.NewCannedPrevOutputFetcher(
		spendingTxOut.PkScript, spendingTxOut.Value,
	)
	sighashes := txscript.NewTxSigHashes(unsignedTx, prevOutputFetcher)
	fakeTapscriptRootHash := []byte{}
	sig, err := txscript.RawTxInTaprootSignature(
		unsignedTx, sighashes, 0, spendingTxOut.Value, spendingTxOut.PkScript,
		fakeTapscriptRootHash, txscript.SigHashDefault, key,
	)
	if err != nil {
		return nil, err
	}

	var signedTxBuf bytes.Buffer
	err = unsignedTx.Serialize(&signedTxBuf)
	if err != nil {
		return nil, err
	}

	signedTxBytes, err := common.UpdateTxWithSignature(signedTxBuf.Bytes(), 0, sig)
	if err != nil {
		return nil, err
	}
	signedTx, err := common.TxFromRawTxBytes(signedTxBytes)
	if err != nil {
		return nil, err
	}

	err = common.VerifySignature(signedTx, 0, spendingTxOut)
	if err != nil {
		return nil, err
	}

	return signedTx, nil
}
